### DISCUSSION TITLE
How to disable modulepreload/prefetch or limit the amount of requests?

### DISCUSSION THREAD

**Discussion by jshimkoski** [nuxt/nuxt#16231] [Questions]
After a production build, over 100 `<link rel="modulepreload">...` and `<link rel="preload">...` tags are injected into the head of the page. They are killing initial page load performance of the app in a major way.

For reference, the Nuxt 2 version of the app can load in about 1 second (with a bundle about 3 times the size of the Nuxt 3 bundle). In Nuxt 3, the same app takes close to 4 seconds to load. I've isolated the issue to the excessive inclusion of these preload `link` elements.

Is there a quick configuration option or some other solution that will disable or limit the amount of these elements that are generated?

**Comment by jshimkoski** [comment]
Just a small update, the only place that I've seen that might allow me to disable the functionality is in `.output/server/node_modules/vue-bundle-renderer/dist/runtime.mjs`, where I have to manually change lines 7 and 8 to set shouldPrefetch and shouldPreload to false:

```
  const ctx = {
    shouldPrefetch: () => false, // shouldPrefetch || defaultShouldPrefetch,
    shouldPreload: () => false, // shouldPreload || defaultShouldPreload,
...
```

Just changing those two lines changes the Lighthouse performance score by 20%, in favor of their removal.

Unfortunately, this is super hacky and not an ideal solution. Still hoping there is a quick configuration option that could be set but nothing seems to work. I've tried this in `nuxt.config.ts` and as far as I can see, it has no effect:

```
  vite: {
    build: {
      modulePreload: {
        polyfill: false,
        resolveDependencies: (filename, deps, { hostId, hostType }) => {
          return []
        }
      },
    },
  },
```

**Reply by sh0ber** [reply]
This is a critically necessary feature to prevent Nuxt 3 from killing performance scores / LCP.  In my testing, Nuxt 3 underperforms a similarly structured Nuxt 2 app despite being smaller and having fewer requests.  It's damaging my Lighthouse scores considerably.

Please give an option to disable preload/prefetch on everything so I don't have to do it manually.

**Comment by Tristan971** [comment]
Following https://github.com/nuxt/nuxt/issues/14584 pretty closely.

While this isn't the literal end of the world, our webapp peak nuxt-assets-related traffic went from 4XX rqps to around 8XXX rqps, which has been a problem for both our users and our infrastructure...

it seems like 3.2.2 builds nearly halved the impact, but it's still quite unfortunate

**Comment by fmoessle** [comment]
Any update on this? We are also experiencing a lot of unnecessary `<link rel="modulepreload">` and `<link rel="preload">` ending up in `<head>`. This is happening with Lazy-components and contradicts their purpose completely.

**Comment by hrehfeld** [comment]
What's the status here? Shouldn't this be quite high-prio?

**Comment by maximilian-schwarz** [comment]
Need this as well

**Comment by bagaskarala** [comment]
i use super hacky approach to disable modulepreload by change the value of `modulepreload` to `nopreload` via `render:html` hook.
it is dirty but it works anyway


update with cleaner approach, add this to nuxt config

```
  hooks: {
    'build:manifest': (manifest) => {
      for (const key in manifest) {
        manifest[key].preload = false;
      }
    },
  },
```

**Reply by andrewspy** [reply]
From my testing, setting `manifest[key].preload = false;` doesn't help as it will still preload in `entry.js`.

**Comment by edwh** [comment]
Lots of versions of this question around, which highlights the general issue of Nuxt and Google's metrics such as LCP.  

IMO that isn't going to go away until Nuxt offers a very clear option for this - even if people think it's silly or counterproductive.  It's the only way to shut us all up :-).

In my case I'm using SSR and have images set to preload, but the page still takes a while to display on slow devices (like mobile) because it's too busy loading and parsing Javascript to render the images.

Here's a hacky module I put in `server\plugins` to strip out the modulepreloads during the build process.

```
export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:response', (response, { event }) => {
    response.body = response.body.replace(
      /<link rel="modulepreload"[\s\S]*?>/g,
      ''
    )
  })
})

```

**Reply by Tristan971** [reply]
As outlined by multiple of us in the likes of #14584, this isn’t just a pagespeed metrics thing. Pagespeed isn’t even particularly relevant for SEO purposes (which tends to be the primary reason people obsess over it) unless you’re talking absurd values in the double-digit seconds.

The number of chunks generated in combination with their needlessly eager preloading, on the other hand, is a very real issue.

It’s just a very inefficient bundling in many ways and fixing it would yield very real benefits (irrelevant of pagespeed metrics).

**Reply by edwh** [reply]
> As outlined by multiple of us in the likes of #14584, this isn’t just a pagespeed metrics thing.

I completely agree, and was on that thread too.  I'm just slightly baffled by why this hasn't been sorted yet with a big "stop preloading" switch which turns it all off.  It's clear that there are enough sane people arguing for it that there is an issue for some of us.

Obviously the team are kindly providing their code for free for all of us, so I don't mean to be ungrateful.  But I can only conclude that at some level the Nuxt folk just don't see this as a significant problem.  

Trying to persuade them isn't working.  I've sometimes implemented stuff which I disagree with just to stop having to have the same conversation over and over.  A big switch would do that.

**Reply by Tristan971** [reply]
Ah yes, I’m also confused why this has been so ostensibly ignored

**Comment by gaisinskii** [comment]
Also experiencing this problem with a lot of ` rel="modulepreload` & `rel="prefetch"` links, any updates on this topic?

**Comment by guska8** [comment]
Nuxt 3 working solution to manipulate/disable prefetch and preload. Posted originally here https://github.com/nuxt/nuxt/issues/13778#issuecomment-2076656036

Add to nuxt.config.ts

```
  hooks: {
    'build:manifest': (manifest) => {
      for (const key in manifest) {
        const file = manifest[key]
        // console.log('-----', file)
        // all file stop prefetch；
        file.preload = false
        file.prefetch = false
      }
    }
  }
```

**Reply by FrankFang** [reply]
It does not work.
